README.md
# Description

Angular animations made easy. [Example Page](https://ackerapple.github.io/ack-angular-fx)

[![hire me](https://ackerapple.github.io/resume/assets/images/hire-me-badge.svg)](https://ackerapple.github.io/resume/)
[![npm version](https://badge.fury.io/js/ack-angular-fx.svg)](http://badge.fury.io/js/ack-angular-fx)
[![npm downloads](https://img.shields.io/npm/dm/ack-angular-fx.svg)](https://npmjs.org/ack-angular-fx)
[![Dependency Status](https://david-dm.org/ackerapple/ack-angular-fx.svg)](https://david-dm.org/ackerapple/ack-angular-fx)

![animated gif](https://ackerapple.github.io/ack-angular-fx/assets/fireworks.gif "gumball fireworks")

> Angular 5.2.0+ required

<details>
  <summary>Table of Contents</summary>

- [Install](#install)
  - [Requirements](#requirements)
- [Include](#include)
- [Browser Compatibility](#browser-compatibility)
- [Examples](#examples)
  - [Robust Example](#robust-example)
  - [Params Example](#params-example)
- [Stagger](#stagger)
- [Supported Animations](#supported-animations)
- [Prebuild Fxs to Single File](#prebuild-fxs-to-single-file)
- [Work on This Project](#work-on-this-project)
  - [Building Example](#building-example)
- [Credits](#credits)
- [Resources](#resources)

</details>

# Install

## Requirements
Since you're using Angular, the following are expected to already be installed:

- reflect-metadata
- @angular/core
- @angular/animations

install command:
```bash
npm install ack-angular-fx --save-dev
```

# Include
Animations must be injected into your components and this is how to do that

```typescript
import { animations } from "ack-angular-fx";
import { Component } from "@angular/core"

@Component({
  animations:animations,
  template:`
    <div *ngIf="boolean" [@fadeInOutUp]="boolean">
      ...
    </div>
    <button (click)="boolean=!boolean">
      hide or show
    </button>
  `
}) export class AppComponent
```


# Browser Compatibility

All good!

Angular animations have come a long way. In the past, they didn't work in Safari without including additional libraries. Now, all animations appear to be working anywhere that Angular could be run.


# Examples

[View interactive example page here.](https://ackerapple.github.io/ack-angular-fx)

## Robust Example
This example has pretty much everything needed to boot

> NOTE: AckFxModule, in code below, is only needed IF directive fx-tracker being used for back-n-forth swap animation tracking

```typescript
import "zone.js"
import "reflect-metadata"
import { platformBrowserDynamic } from "@angular/platform-browser-dynamic"
import { NgModule, Component } from "@angular/core"
import { BrowserModule } from "@angular/platform-browser";
import { BrowserAnimationsModule } from "@angular/platform-browser/animations"

import { animations } from "ack-angular-fx";

//NOTE: This line only needed IF directive fx-tracker used (back-n-forth swap animations)
import { AckFxModule } from "ack-angular-fx";
  
import { supportDocument } from "ack-angular-fx/web-animations.min"

supportDocument( document )//cross browser fx support

@Component({
  animations:animations,
  template:`
    <div *ngIf="boolean" [@rotateInOut]="boolean">
      ...
    </div>
    <button (click)="boolean=!boolean">
      hide or show
    </button>
  `
}) export class AppComponent

@NgModule({
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    AckFxModule//NOTE: This line only needed IF directive fx-tracker used (back-n-forth swap animations)
  ],
  declarations: [ AppComponent ],
  bootstrap: [ AppComponent ]
}) export class AppModule {}


platformBrowserDynamic().bootstrapModule(AppModule)
```

## Params Example
The configuration options available to define for animations

```html
<div [@zoomInOut]="{value:boolean, params:{time:'2000ms 0 linear'}}">
  ...
</div>
```

Learn more about [Angular animation timing here](https://angular.io/guide/animations#animation-timing)

# Stagger
Offset multiple animations using Angular 4.2.4 or greater

Currently, staggering animations using ack-angular-fx, is controlled by a parent stagger definition and child animations subscribing to that via the css class "childFx"

#### Stagger Items Example
A parent container staggering child animations whenever the array adds or removes elements
```html
<div [@childStag]="array.length">
  <div
    *ngFor = "let x of array"
    [@rotateInOut] = "1"
  >
    <div>
      ngFor loop with animations for each
    </div>
  </div>
</div>
```

#### Stagger Table Rows Example
A table element staggering row animations whenever "rowStaggers" variable changes value
```html
<table style="width:100%" [@childStag]="array.length">
  <ng-container *ngFor="let x of array">
    <tr [@bounceInOut]="array.length" style="background-color:#CCC">
      <td>{{ x }}</td>
      <td>Item Row {{ x }}</td>
    </tr>
  </ng-container>
</table>
```

# Supported Animations

- fadeIn
  - fadeInDown, fadeInLeft, fadeInRight, fadeInUp
- fadeOut
  - fadeOutDown, fadeOutLeft, fadeOutRight, fadeOutUp
- fadeInOut
  - fadeInOutDown, fadeInOutLeft, fadeInOutRight, fadeInOutUp
- bounceIn
  - bounceInDown, bounceInLeft, bounceInRight, bounceInUp
- bounceOut
  - bounceOutDown, bounceOutLeft, bounceOutRight, bounceOutUp
- bounceIn
  - bounceInOutDown, bounceInOutLeft, bounceInOutRight, bounceInOutUp
- rotateIn
  - rotateInDownLeft, rotateInDownRight, rotateInUpLeft, rotateInUpRight
- rotateOut
  - rotateOutDownLeft, rotateOutDownRight, rotateOutUpLeft, rotateOutUpRight
- rotateInOut
  - rotateInOutDownLeft, rotateInOutDownRight, rotateInOutUpLeft, rotateInOutUpRight
- zoomIn
  - zoomInDown, zoomInLeft, zoomInRight, zoomInUp
- zoomOut
  - zoomOutDown, zoomOutLeft, zoomOutRight, zoomOutUp
- zoomInOut
  - zoomInOutDown, zoomInOutLeft, zoomInOutRight, zoomInOutUp
- slideInDown, slideInLeft, slideInRight, slideInUp
- slideOutDown, slideOutLeft, slideOutRight, slideOutUp
- slideInOutDown, slideInOutLeft, slideInOutRight, slideInOutUp

see online demo https://ackerapple.github.io/ack-angular-fx

## Prebuild Fxs to Single File
Known as the prefx process, ack-angular-fx has cli commands to bundle animations to file

> This often helps with bundling issues and Angular AOT compilation

### Example Compile All Animations to One File

Step 1 : Add the following to your package.json scripts
```javascript
"scripts":{
  "build:prefx": "ack-angular-fx --out ./src/prefx.ts"
}
```

Step 2 : Now, run the following in a command prompt terminal
```bash
npm run build:prefx
```
> After the above command you should now see a prefx.ts file in your src folder

Step 3 : Import your compiled .ts file and apply to component(s)
```javascript
import { animations } from "./prefx"
import { Component } from "@angular/core"

@Component({
  selector: 'app',
  template: '<div *ngIf="a==1" [@fadeInUp]="true">hello world</div>',
  animations: animations
})
```


# Work on This Project
Everything in this topic is run in an command prompt terminal

Clone this project
```bash
git clone https://github.com/AckerApple/ack-angular-fx
```

Dev Fast Project Watching
```bash
npm run watch
```

Build before commit
```bash
npm run build
```


# Credits

A special thank you goes out to [yuyang041060120](https://github.com/yuyang041060120/) for creating [ng2-animate](https://github.com/yuyang041060120/ng2-animate) and for taking too long to respond to my [pull request](https://github.com/yuyang041060120/ng2-animate/pull/4), which has led to the birth of ack-angular-fx.

[Table of Contents](#table-of-contents)

# Resources
- Inspired by [Animate.css](https://daneden.github.io/animate.css)
- [tester @ matthewlein.com](https://matthewlein.com/ceaser/)
- [maps @ easings.net](http://easings.net/)


CHANGELOG.md
# ack-angular-fx - Change Log
All notable changes to this project will be documented here.

## [4.2.0] - 2022-01-04
- Add peerDependencies

## [3.0.4] - 2019-05-26
- Fix tracking history by orderArray

## [3.0.3] - 2019-05-18
- better fx-tracker history

## [3.0.2] - 2019-04-17
- Added fx-tracker inFx helper

## [3.0.1] - 2019-03-19
- fx-tracker uses Promise.resolve().then(()=>) instead of setTimeout

## [3.0.0] - 2019-02-13
- Added Out
  - fadeOut
- Updated dependencies
- Breaking Changes
  - Removed "Global Usage" injection that came via 
  - all in Kids animations like fadeInKids is renamed to fadeInOutKids
  - In is now InOut
    - fadeIn is now fadeInOut
    - fadeIn only transistions in but not out anymore
  - "fxArray" is renamed to "animations"
    - `import { animations } from "ack-angular-fx"`

## [2.3.0] - 2018-11-16
- Updated Angular

## [2.2.1] - 2018-09-28
- fix file refs

## [2.2.0] - 2018-09-28
- Upgraded to Angular6
- deprecated fxId and fxIdChange
  - use `<fx-tracker #fx="FxTracker"></fx-tracker>`
  - fx.id
- removed web-animations-js

## [2.1.2] - 2018-04-12
- Made all basic animations animate even when false or 0 are based in
  - Condition to check to animate was ignoring false or 0
  - Pass 'nofx' when you dont want animation

## [2.1.1] - 2018-03-03
- Added router-outlet documentation
- Fixed FxTracker value tracking when activeRouter monitoring

## [2.1.0] - 2018-03-01
- improved file size by deprecating Angular4 animations
- Breaking Changes
  - Removed [@100],[@200],[@300],[@400],[@500],...
  - Removed [@absoluteSwap],[@absoluteSwap100],[@absoluteSwap200],...

## [2.0.11] - 2018-03-01
- allFx file is now a prefx file and apart of the ack-angular-fx process

## [2.0.9] - 2018-03-01
- all animation transition functions are exported

## [2.0.8] - 2018-02-28
- Updated README file

## [2.0.6] - 2018-02-23
- added transistion function interpolation
  - with options compiled as well
- fix rotate animations

## [2.0.0] - 2018-02-22
- prefx compiling seems no longer required
### Breaking Changes
- upgraded to Angular5. Angular5 is required to use this package
- old stuff should still work but is becoming deprecated

## [1.2.5] - 2017-07-06
### Fixed
- user reported issue
### Changed
- 500 max to 300 for childStag

## [1.2.0] - 2017-06-29
### Breaking Changes
- Angular 4.2.4 is now required
### Added
- query, stagger, animateChild support
- new childStag trigger to support staggering

## [1.1.0] - 2017-03-30
### Fixed
- Made more AoT compatible by making AnimationTriggerMetadata available to prefx outputs

## [1.0.11] - 2017-03-20
### Added
- Better docs

## [1.0.6] - 2017-03-16
### Added
- AoT support by making static prefx animations

## [1.0.5] - 2017-03-09


src/helpers.ts
import { animations } from "./allFxDynamic"
export { animations } from "./allFxDynamic"

import {
  selectedFxMetaData, fxConfig, effectsTypeObject
} from "./fxTypes"

//import { states as fadeStates } from './animations/fade';
import { bounce } from './animations/bounce';
import { rotate } from './animations/rotate';
import { slide } from './animations/slide';
import { zoom } from './animations/zoom';

import {
  trigger,
  AnimationTriggerMetadata,
  AnimationStateMetadata,
  AnimationTransitionMetadata
} from '@angular/animations';

export const availEffects:string[] = ['fade','bounce','rotate','slide','zoom']
export const delayArray:number[] = [100,200,300,400,500,600,700,800,900,1000,1500,2000]
export const animateDefaults = {
  //duration   : 500,
  //delay      : 0,
  //easing     : 'linear',
  stagger    : 0,
  name       : 'animate',
  igniter    : '*',
  whileStyle : {},
  effects    : availEffects
}

export const absSwap = {
  easing:'linear', name:'absoluteSwap', igniter:'void',
  whileStyle:{
    position: 'absolute', width:'100%', 'overflow':'hidden'
  }
}

export const menu = {
  absoluteSwap:{duration:500, ...absSwap},
  absoluteSwap100:{duration:100, ...absSwap},
  absoluteSwap200:{duration:200, ...absSwap},
  absoluteSwap300:{duration:300, ...absSwap},
  absoluteSwap400:{duration:400, ...absSwap},
  absoluteSwap500:{duration:500, ...absSwap},
  absoluteSwap600:{duration:600, ...absSwap},
  absoluteSwap700:{duration:700, ...absSwap},
  absoluteSwap800:{duration:800, ...absSwap},
  absoluteSwap900:{duration:900, ...absSwap},
  absoluteSwap1000:{duration:1000, ...absSwap},
  absoluteSwap1500:{duration:1500, ...absSwap},
  absoluteSwap2000:{duration:2000, ...absSwap},
  absoluteSwap2500:{duration:2500, ...absSwap},

  "100":{duration:100},
  "200":{duration:200},
  "300":{duration:300},
  "400":{duration:400},
  "500":{duration:500},
  "600":{duration:600},
  "700":{duration:700},
  "800":{duration:800},
  "900":{duration:900},
  "1000":{duration:1000},
  "1500":{duration:1500},
  "2000":{duration:2000}
}

export function getConfigTiming(
  config:fxConfig
):string{
  return [
    (config.duration || 500) +'ms',
    (config.delay || 0) + 'ms',
    (config.easing  || 'linear')
  ].join(' ')
}


export function defaultConfig(config){
  return {
    ...animateDefaults,
    ...config
  }
}

/*export function checkStagger(config){
  return (!config.stagger || console.log('ack-angular-fx does not support stagger as of this release')) && config
}*/

export function animateConfig(
  name:string,
  config:fxConfig
) : AnimationTriggerMetadata{
  const dConfig = defaultConfig(config)
  const timing = getConfigTiming(dConfig)
  return createTriggerBy(name, config, timing)
}

export function createTriggerBy(
  name:string,
  config:fxConfig,
  timing:string
) : AnimationTriggerMetadata{
  return trigger(name, stateEffectsByConfig(timing,config))
}

export function effectsArrayToTypes( eArray:string[] ) : effectsTypeObject {
  const ob = {}

  for(let x=availEffects.length-1; x >= 0; --x){
    let fx = availEffects[x]
    if( eArray.indexOf(fx)>=0 ){
      ob[fx] = true
    }
  }

  return ob
}

export function stateEffectsByConfig(
  timing:string,
  config:fxConfig
) : (AnimationStateMetadata|AnimationTransitionMetadata)[] {
  const array = []
  const fxTypes = effectsArrayToTypes( config.effects || availEffects )

  /*if( fxTypes.fade ){
    array.push(...fadeStates(config))
  }*/

  if( fxTypes.bounce ){
    array.push(...bounce(timing, config))
  }

  if( fxTypes.rotate ){
    array.push(...rotate(timing, config))
  }

  if( fxTypes.slide ){
    array.push(...slide(timing, config))
  }

  if( fxTypes.zoom ){
    array.push(...zoom(timing, config))
  }

  return array
}

export function selectFx(
  args:string[],
  effectList:string[],
  config = <fxConfig>{igniter:'start'}
) : selectedFxMetaData {
  const selectedFx = { states:[], triggers:[] }

  //loop an array like [100,200,"childStag"]
  args.forEach(fxType=>{
    const cloneConfig:fxConfig = {...menu[fxType]}
    const newConfig = {...cloneConfig, ...config}
    const newSelect = processSelect(
      fxType,
      newConfig
      //,effectList
    )
    selectedFx.states.push( newSelect )
  })

  selectedFx.triggers = processTriggerSelect(config, effectList)

  return selectedFx
}

/** Goes into specialized fx files and plucks out the stand alone triggers */
// ---- NOT DONE ----
function processTriggerSelect(
  _config:fxConfig,
  _effectList
){
  const fxs = []
  //const fxTypes = effectsArrayToTypes( config.effects || availEffects )
  
  fxs.push.apply(fxs, animations)
  
  return fxs
}

export function processSelect(
  name:string,
  config:fxConfig
  //,effectArray?:string[]
) : AnimationTriggerMetadata{
  config.igniter = config.igniter || 'void'
  return animateConfig(name, config)
}

/*export const easeArray = [
  {name:'Ease', value:'ease'},
  {name:'EaseIn', value:'ease-in'},
  {name:'EaseOut', value:'ease-out'},
  {name:'EaseInOut', value:'ease-in-out'}
]*/


//animations + deprecated fxs
export function getAnimations() : AnimationTriggerMetadata[] {
  const allFx = [
    processSelect("absoluteSwap", menu["absoluteSwap400"] ),
    processSelect("absoluteSwap100", menu["absoluteSwap100"] ),
    processSelect("absoluteSwap200", menu["absoluteSwap200"] ),
    processSelect("absoluteSwap300", menu["absoluteSwap300"] ),
    processSelect("absoluteSwap400", menu["absoluteSwap400"] ),
    processSelect("absoluteSwap500", menu["absoluteSwap500"] ),
    processSelect("absoluteSwap600", menu["absoluteSwap600"] ),
    processSelect("absoluteSwap700", menu["absoluteSwap700"] ),
    processSelect("absoluteSwap800", menu["absoluteSwap800"] ),
    processSelect("absoluteSwap900", menu["absoluteSwap900"] ),
    processSelect("absoluteSwap1000", menu["absoluteSwap1000"] ),
    processSelect("absoluteSwap1500", menu["absoluteSwap1500"] ),
    processSelect("absoluteSwap2000", menu["absoluteSwap2000"] ),
    processSelect("100", menu["100"] ),
    processSelect("200", menu["200"] ),
    processSelect("300", menu["300"] ),
    processSelect("400", menu["400"] ),
    processSelect("500", menu["500"] ),
    processSelect("600", menu["600"] ),
    processSelect("700", menu["700"] ),
    processSelect("800", menu["800"] ),
    processSelect("900", menu["900"] ),
    processSelect("1000", menu["1000"] ),
    processSelect("1500", menu["1500"] ),
    processSelect("2000", menu["2000"] )
  ]

  allFx.push.apply(allFx, animations)

  return allFx
}

export let absSwapClone = {
  name:null, duration:null, whileStyle:null
}